runtime/compiler/control/CompilationThread.cpp

```diff
@@ -9294,6 +9331,390 @@ TR::CompilationInfoPerThreadBase::performAOTLoad(
+      trCountFromMJIT = (trCountFromMJIT << 1) | 1;\
+      TR_J9VMBase *fe = TR_J9VMBase::get(jitConfig, vmThread);\
+      U_8 *extendedFlags = fe->fetchMethodExtendedFlagsPointer(method);\
+      *extendedFlags = (*extendedFlags) | J9_MJIT_FAILED_COMPILE;\
+      method->extra = reinterpret_cast<void *>(trCountFromMJIT);\
+      compiler->failCompilation<MJIT::MJITCompilationFailure>("Cannot compile with MicroJIT.");\
+      }\
+} while(0)
+
+// This routine should only be called from wrappedCompile
+TR_MethodMetaData *
+TR::CompilationInfoPerThreadBase::mjit(
+   J9VMThread *vmThread,
+   TR::Compilation *compiler,
+   TR_ResolvedMethod *compilee,
+   TR_J9VMBase &vm,
+   TR_OptimizationPlan *optimizationPlan,
+   TR::SegmentAllocator const &scratchSegmentProvider,
+   TR_Memory *trMemory
+   )
+   {
+
+   PORT_ACCESS_FROM_JITCONFIG(jitConfig);
+
+   TR_MethodMetaData *metaData = NULL;
+   J9Method *method = NULL;
+   TR::CodeCache *codeCache  = NULL;
+
+   if (_methodBeingCompiled->_priority >= CP_SYNC_MIN)
+      ++_compInfo._numSyncCompilations;
+   else
+      ++_compInfo._numAsyncCompilations;
+
+   if (_methodBeingCompiled->isDLTCompile())
+      compiler->setDltBcIndex(static_cast<J9::MethodInProgressDetails &>(_methodBeingCompiled->getMethodDetails()).getByteCodeIndex());
+
+   bool volatile haveLockedClassUnloadMonitor = false; // used for compilation without VM access
+   int32_t estimated_size = 0;
+   char* buffer = NULL;
+   try
+      {
+
+      InterruptibleOperation compilingMethodBody(*this);
+
+      TR::IlGeneratorMethodDetails & details = _methodBeingCompiled->getMethodDetails();
+      method = details.getMethod();
+
+      TR_J9VMBase *fe = TR_J9VMBase::get(jitConfig, vmThread); //TODO: fe can be replaced by vm and removed from here.
+      U_8 *extendedFlags = fe->fetchMethodExtendedFlagsPointer(method);
+      const char* signature = compilee->signature(compiler->trMemory());
+
+      TRIGGER_J9HOOK_JIT_COMPILING_START(_jitConfig->hookInterface, vmThread, method);
+
+      // BEGIN MICROJIT
+      TR::FilePointer *logFileFP = this->getCompilation()->getOutFile();
+      MJIT::ByteCodeIterator bcIterator(0, static_cast<TR_ResolvedJ9Method *> (compilee), static_cast<TR_J9VMBase *> (&vm), comp());
+
+      if (TR::Options::canJITCompile())
+         {
+         TR::Options *options = TR::Options::getJITCmdLineOptions();
+
+         J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+         U_16 maxLength = J9UTF8_LENGTH(J9ROMMETHOD_SIGNATURE(romMethod));
+         if(compiler->getOption(TR_TraceCG)){
+            trfprintf(logFileFP, "\n------------------------------\nMicroJIT Compiling %s...\n------------------------------\n", signature);
+         }
+
+         MJIT_COMPILE_ERROR(!(compilee->isConstructor()), method, romMethod); //"MicroJIT does not currently compile constructors"
```

Comment:
Rather than throwing an exception here, it's better to test for constructor at the point where we decide whether to compile with microjit or with TR jit